base_multi_image/models/image.py

```diff
@@ -0,0 +1,217 @@
+        self.show_technical = all(
+            "default_owner_%s" % f not in self.env.context
+            for f in ("id", "model"))
+
+    @api.multi
+    def _get_image_from_filestore(self):
+        return self.attachment_id.datas
+
+    @api.multi
+    def _get_image_from_db(self):
+        return self.file_db_store
+
+    @api.multi
+    def _get_image_from_file(self):
+        if self.path and os.path.exists(self.path):
+            try:
+                with open(self.path, 'rb') as f:
+                    return base64.b64encode(f.read())
+            except Exception as e:
+                _logger.error("Can not open the image %s, error : %s",
+                              self.path, e, exc_info=True)
+        else:
+            _logger.error("The image %s doesn't exist ", self.path)
+
+        return False
+
+    @api.multi
+    def _get_image_from_url(self):
+        return self._get_image_from_url_cached(self.url)
+
+    @api.model
+    @tools.ormcache("url")
+    def _get_image_from_url_cached(self, url):
+        """Allow to download an image and cache it by its URL."""
+        if url:
+            try:
+                (filename, header) = urllib.urlretrieve(url)
+                with open(filename, 'rb') as f:
+                    return base64.b64encode(f.read())
+            except:
+                _logger.error("URL %s cannot be fetched", url,
+                              exc_info=True)
+
+        return False
+
+    @api.multi
+    @api.depends('image_main')
+    def _get_image_sizes(self):
+        for s in self:
+            try:
+                vals = tools.image_get_resized_images(
+                    s.with_context(bin_size=False).image_main)
+            except:
+                vals = {"image_medium": False,
+                        "image_small": False}
+            s.update(vals)
+
+    @api.model
+    def _make_name_pretty(self, name):
+        return name.replace('_', ' ').capitalize()
+
+    @api.onchange('url')
+    def _onchange_url(self):
+        if self.url:
+            filename = self.url.split('/')[-1]
+            self.name, self.extension = os.path.splitext(filename)
+            self.name = self._make_name_pretty(self.name)
+
+    @api.onchange('path')
+    def _onchange_path(self):
+        if self.path:
+            self.name, self.extension = os.path.splitext(os.path.basename(
+                self.path))
+            self.name = self._make_name_pretty(self.name)
+
+    @api.onchange('filename')
+    def _onchange_filename(self):
+        if self.filename:
+            self.name, self.extension = os.path.splitext(self.filename)
+            self.name = self._make_name_pretty(self.name)
+
+    @api.onchange('attachment_id')
+    def _onchange_attachmend_id(self):
+        if self.attachment_id:
+            self.name = self.attachment_id.res_name
+
+    @api.constrains('storage', 'url')
+    def _check_url(self):
+        if self.storage == 'url' and not self.url:
```

Comment:
When user will add multiple images at a time from any kanban / one2many view, this will raise error of "ValueError: expected singleton ". So self object needs to be traversed